proposals/0048-precompile-for-secp256r1-sigverify.md

```diff
@@ -0,0 +1,274 @@
+data with a protocol that only accepts compressed points.
+
+**Note:** When it comes to public key encoding, the existing
+precompile for `secp256k1` utilizes a vastly different standard,
+accepting a 20 octet Ethereum address and recovery id to recover an
+SEC1 encoded uncompressed point. This is due to the primary aim of the
+program not being to verify ECDSA signatures but to provide parity with
+`ecrecover` on EVM. Conversely, the `ed25519` program, which is
+primarily concerned with verifying ed25519 signatures, utilises the most
+common ed25519 convention of encoding x_p as a single 32 octet string.
+As the goals of the `secp256r1` program are more analogous to those of the
+`ed25519` program, we propose the SEC1 compressed point encoding to conform
+to the most widely-used standard in common ECDSA applications.
+
+### ECDSA / Signature Verification
+
+The precompile should implement the `Verifying Operation` outlined in
+[SEC1](https://www.secg.org/sec1-v2.pdf#page=52)
+in Section 4.1.4 as well as in the
+[Digital Signature Standard (DSS)](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf#page=36)
+document in Section 6.4.2.
+
+A multitude of test vectors to verify correctness can
+be found in
+[RFC6979](https://datatracker.ietf.org/doc/html/rfc6979#appendix-A.2.5)
+in Section A.2.5 as well as at the
+[NIST CAVP](https://csrc.nist.gov/Projects/cryptographic-algorithm-validation-program/digital-signatures#ecdsa2vs)
+(Cryptographic Algorithm Validation Program)
+
+### Signature Malleability
+
+Due to X axis symmetry along the elliptic curve, for any ECDSA signature
+`(r, s)`, there also exists a valid signature `(r, n - s)`, where `n` is the
+order of the curve. This introduces "s malleability", allowing an attacker
+to produce an alternative version of `s` without invalidating the signature.
+
+The pitfalls of this in authentication systems can be particularly perilous,
+opening up certain implementations to signature replay attacks over the same
+message by simply flipping the `s` value over the curve.
+
+As the primary goal of the `secp256r1` program is secure signature validation
+for authentication purposes, the precompile should mitigate these attacks
+by enforcing the usage of `lowS` values, in which `s <= n/2`.
+
+As such, the program should immediately fail upon the detection of any
+signature that includes a `highS` value. This prevents any accidental
+succeptibility to signature malleability attacks.
+
+Note: The existing `secp256k1` precompile makes no attempt attempt to mitigate
+s malleability, as doing so would go against its primary goal of achieving
+`ecrecover` parity with EVM.
+
+### Program
+
+ID: `Secp256r1SigVerify1111111111111111111111111`
+
+The program instruction will be composed of the following:
+
+- A first u8 as the count for the number of signatures to check
+- Single byte of padding
+- The following struct serialized, for each signature to verify
```

Comment:
please specify this layout as a pseudo-coded struct. avoid assuming implementation language, be explicit about things like alignment, padding, signedness, endianness, bitwidth, (dis)allowed enum values, etc. idea is that someone can read this doc and implement it from scratch in whatever language they like with no other references

applies to the struct below as well